diff --git openSSL.orig/crypto/CMakeLists.txt OpenSSL.new/crypto/CMakeLists.txt
index 2aba30a..bb7a5bb 100644
--- openSSL.orig/crypto/CMakeLists.txt
+++ OpenSSL.new/crypto/CMakeLists.txt
@@ -278,7 +278,7 @@ else()
   add_submodule ( . threads_pthread.c )
 endif()
 
-add_library( crypto ${LIBSRC} ${OBJECTS_SRC} )
+add_library( crypto SHARED ${LIBSRC} ${OBJECTS_SRC} )
 
 target_include_directories( crypto PUBLIC ${PROJECT_BINARY_DIR}/include )
 
@@ -297,7 +297,7 @@ set_target_properties( crypto PROPERTIES
   VERSION "${LIB_VERSION}" SOVERSION "${LIB_SOVERSION}" )
 
 if( ANDROID )
-  set_target_properties( crypto PROPERTIES OUTPUT_NAME "crypto_1_1" )
+  set_target_properties( crypto PROPERTIES OUTPUT_NAME "crypto" )
 endif()
 
 if( MSVC )
diff --git openSSL.orig/opensslconf.h.cmake OpenSSL.new/opensslconf.h.cmake
index c5a5133..09297d6 100644
--- openSSL.orig/opensslconf.h.cmake
+++ OpenSSL.new/opensslconf.h.cmake
@@ -51,9 +51,6 @@ extern "C" {
 #ifndef OPENSSL_NO_EC_NISTP_64_GCC_128
 # define OPENSSL_NO_EC_NISTP_64_GCC_128
 #endif
-#ifndef OPENSSL_NO_EGD
-# define OPENSSL_NO_EGD
-#endif
 #ifndef OPENSSL_NO_EXTERNAL_TESTS
 # define OPENSSL_NO_EXTERNAL_TESTS
 #endif
@@ -75,12 +72,6 @@ extern "C" {
 #ifndef OPENSSL_NO_SSL_TRACE
 # define OPENSSL_NO_SSL_TRACE
 #endif
-#ifndef OPENSSL_NO_SSL3
-# define OPENSSL_NO_SSL3
-#endif
-#ifndef OPENSSL_NO_SSL3_METHOD
-# define OPENSSL_NO_SSL3_METHOD
-#endif
 #ifndef OPENSSL_NO_UBSAN
 # define OPENSSL_NO_UBSAN
 #endif
diff --git openSSL.orig/ssl/CMakeLists.txt OpenSSL.new/ssl/CMakeLists.txt
index a1ff744..66257c9 100644
--- openSSL.orig/ssl/CMakeLists.txt
+++ OpenSSL.new/ssl/CMakeLists.txt
@@ -43,7 +43,7 @@ include_directories( BEFORE SYSTEM
   ${PROJECT_SOURCE_DIR}/ # e_os.h
 )
 
-add_library( ssl ${LIBSRC} )
+add_library( ssl SHARED ${LIBSRC} )
 
 target_include_directories( ssl PUBLIC ${PROJECT_BINARY_DIR}/include )
 
@@ -53,7 +53,7 @@ set_target_properties( ssl PROPERTIES
   VERSION "${LIB_VERSION}" SOVERSION "${LIB_SOVERSION}" )
 
 if( ANDROID )
-  set_target_properties( ssl PROPERTIES OUTPUT_NAME "ssl_1_1" )
+  set_target_properties( ssl PROPERTIES OUTPUT_NAME "ssl" )
 endif()
 
 if( MSVC )
