diff --git SDL2_gpu.orig/CMakeLists.txt SDL2_gpu.new/CMakeLists.txt
index 6a8f7e4..bcdf40b 100644
--- SDL2_gpu.orig/CMakeLists.txt
+++ SDL2_gpu.new/CMakeLists.txt
@@ -289,8 +289,10 @@ else (DISABLE_GLES)
     
     include_directories(src/externals/gl3stub)
     if (DYNAMIC_GLES_3)
-        SET(SDL_gpu_HDRS ${SDL_gpu_HDRS} externals/gl3stub/gl3stub.c externals/gl3stub/gl3stub.h)
+        SET(SDL_gpu_HDRS ${SDL_gpu_HDRS} externals/gl3stub/gl3stub.h)
         add_definitions("-DSDL_GPU_DYNAMIC_GLES_3")
+		find_library(EGL_LIBRARY EGL)
+		list(APPEND GL_LIBRARIES ${EGL_LIBRARY} )
     endif(DYNAMIC_GLES_3)
 endif (DISABLE_GLES)
 
diff --git SDL2_gpu.orig/include/SDL_gpu_GLES_1.h SDL2_gpu.new/include/SDL_gpu_GLES_1.h
index bca7efe..ba9b09a 100644
--- SDL2_gpu.orig/include/SDL_gpu_GLES_1.h
+++ SDL2_gpu.new/include/SDL_gpu_GLES_1.h
@@ -5,7 +5,7 @@
 #include "SDL_platform.h"
 
 #if !defined(SDL_GPU_DISABLE_GLES) && !defined(SDL_GPU_DISABLE_GLES_1)
-
+#define GL_GLEXT_PROTOTYPES
 #ifdef __IPHONEOS__
     #include <OpenGLES/ES1/gl.h>
     #include <OpenGLES/ES1/glext.h>
@@ -13,7 +13,7 @@
     #include "GLES/gl.h"
     #include "GLES/glext.h"
 #endif
-
+#undef GL_GLEXT_PROTOTYPES
     #define glFrustum glFrustumf
     #define glOrtho glOrthof
     #define glGenerateMipmap glGenerateMipmapOES
diff --git SDL2_gpu.orig/src/CMakeLists.txt SDL2_gpu.new/src/CMakeLists.txt
index 42b5818..f69d26e 100644
--- SDL2_gpu.orig/src/CMakeLists.txt
+++ SDL2_gpu.new/src/CMakeLists.txt
@@ -153,9 +153,12 @@ if(BUILD_SHARED)
 				OUTPUT_NAME ${SHARED_OUTPUT}
 				CLEAN_DIRECT_OUTPUT 1
 				ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/${OUTPUT_DIR}/lib"
-				LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/${OUTPUT_DIR}/lib"
 				RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/${OUTPUT_DIR}/bin"
 			)
+			if(NOT DEFINED CMAKE_LIBRARY_OUTPUT_DIRECTORY)
+				set_target_properties(SDL_gpu_shared PROPERTIES
+					LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/${OUTPUT_DIR}/lib")
+			endif()
 		else()
 			set_target_properties(SDL_gpu_shared PROPERTIES
 				OUTPUT_NAME ${SHARED_OUTPUT}
diff --git SDL2_gpu.orig/src/SDL_gpu.c SDL2_gpu.new/src/SDL_gpu.c
index 8ff0151..51b77c9 100644
--- SDL2_gpu.orig/src/SDL_gpu.c
+++ SDL2_gpu.new/src/SDL_gpu.c
@@ -427,8 +427,13 @@ GPU_Target* GPU_Init(Uint16 w, Uint16 h, GPU_WindowFlagEnum SDL_flags)
     for(i = 0; i < renderer_order_size; i++)
     {
         GPU_Target* screen = GPU_InitRendererByID(renderer_order[i], w, h, SDL_flags);
-        if(screen != NULL)
+        if (screen != NULL) {
             return screen;
+        } else {
+            // If we cant initialize this renderer, ignore the errors and try the next one.
+            (void) GPU_PopErrorCode();
+            (void) GPU_PopErrorCode();
+        }
     }
 
     GPU_PushErrorCode("GPU_Init", GPU_ERROR_BACKEND_ERROR, "No renderer out of %d was able to initialize properly", renderer_order_size);
diff --git SDL2_gpu.orig/src/externals/gl3stub/gl3stub.c SDL2_gpu.new/src/externals/gl3stub/gl3stub.c
index d918b2c..807d311 100644
--- SDL2_gpu.orig/src/externals/gl3stub/gl3stub.c
+++ SDL2_gpu.new/src/externals/gl3stub/gl3stub.c
@@ -17,6 +17,10 @@
 #include <EGL/egl.h>
 #include "gl3stub.h"
 
+#ifdef glUnmapBuffer
+#  undef glUnmapBuffer
+#endif
+
 GLboolean gl3stubInit() {
 #define FIND_PROC(s) s = (void*)eglGetProcAddress(#s);
   FIND_PROC(glReadBuffer);
