--- SDL2_mixer.orig/external/libmikmod-3.1.12/Android.mk
+++ SDL2_mixer.new/external/libmikmod-3.1.12/Android.mk
@@ -63,4 +63,4 @@ LOCAL_LDLIBS :=
 LOCAL_STATIC_LIBRARIES :=
 LOCAL_SHARED_LIBRARIES :=
 
-include $(BUILD_SHARED_LIBRARY)
+include $(BUILD_STATIC_LIBRARY)
--- SDL2_mixer.orig/external/smpeg2-2.0.0/Android.mk
+++ SDL2_mixer.new/external/smpeg2-2.0.0/Android.mk
@@ -41,4 +41,4 @@ LOCAL_SHARED_LIBRARIES := SDL2
 
 LOCAL_EXPORT_C_INCLUDES += $(LOCAL_C_INCLUDES)
 
-include $(BUILD_SHARED_LIBRARY)
+include $(BUILD_STATIC_LIBRARY)
